DOCS-1114: Document how to flash an ESP32 with Micro-RDK installer CLI (#2024)
viamrobotics/docs: docs/installation/prepare/microcontrollers/_index.md

description: "Set up the Espressif ESP32 with the micro-RDK."
aliases:
  - /installation/microcontrollers

> # SMEs: Nicolas M., Gautham V.

Reviewer: add Andrew. M as well, I think?

Author: Sounds good, done!

With this installation, you can use your ESP32 with all supported resource APIs, but you cannot write your own code directly interacting with the chip.
If you want to program the chip directly, follow the setup instructions in [the Micro-RDK Development Setup](/installation/prepare/microcontrollers/development-setup/) instead.

### Download the Micro-RDK Installer

Select the appropriate pre-built binary for the architecture of your development machine:

- [Linux (x86_64)](https://github.com/viamrobotics/micro-rdk/releases/latest/download/micro-rdk-installer-amd64-linux)
- [Linux (Aarch64)](https://github.com/viamrobotics/micro-rdk/releases/latest/download/micro-rdk-installer-arm64-linux)
- [MacOS](https://github.com/viamrobotics/micro-rdk/releases/latest/download/micro-rdk-installer-macos)
- [Windows](https://github.com/viamrobotics/micro-rdk/releases/latest/download/micro-rdk-installer-windows.exe)

### Flash your ESP32 with the Micro RDK Installer

Navigate to [the Viam app](https://app.viam.com) and [add a new robot](/manage/fleet/robots/#add-a-new-robot) in your desired location.

1. Click on the name of the robot to go to the robot's page.
2. Click on the **Setup** tab.

> 3. Regardless of your operating system, select **Mac** and press the button that appears in Step 2 to download the Viam app config for your robot.

Reviewer: Q:  Could we change these instructions once https://viam.atlassian.net/browse/RSDK-5080 is merged?

Author: Can do, I made a ticket to track: https://viam.atlassian.net/jira/software/c/projects/DOCS/boards/31?quickFilter=41&selectedIssue=DOCS-1273

   In your terminal, `cd` to the directory where you downloaded your pre-built binary and run the following command to flash the micro-RDK directly to an ESP32 connected to your computer through a data cable:

   {{< tabs name="Download pre-built binaries" >}}
   {{% tab name="Linux (x86_64)"%}}

```sh { class="command-line" data-prompt="$"}
./micro-rdk-installer-amd64-linux write-flash --app-config=<your-file-path-to/viam.json>
```

    {{% /tab %}}
    {{% tab name="Linux (Aarch64)" %}}

```sh { class="command-line" data-prompt="$"}
./micro-rdk-installer-arm64-linux write-flash --app-config=<your-file-path-to/viam.json>
```

    {{% /tab %}}
    {{% tab name="MacOS" %}}

```sh { class="command-line" data-prompt="$"}
./micro-rdk-installer-macos write-flash --app-config=<your-file-path-to/viam.json>
```

    {{% /tab %}}
    {{% tab name="Windows" %}}

```sh { class="command-line" data-prompt="$"}

> ./micro-rdk-installer-windows write-flash --app-config=<your-file-path-to/viam.json>

Reviewer: `./micro-rdk-installer-windows.exe`. That file suffix is Windows specific

Author: Sounds good, done

- **Minimal configuration:** 384kB Ram 4MB flash
- **Recommended configuration:** 384kB Ram + 8MB SPIRAM + 4MB Flash

## Install the Micro-RDK

> Viam provides the [Micro-RDK Installer](https://github.com/viamrobotics/micro-rdk/tree/main/micro-rdk-installer), a CLI that allows a user to flash a build of Micro-RDK, along with their robot's credentials and their wifi information, directly to their ESP32.

Author: ```suggestion
The [Micro-RDK Installer](https://github.com/viamrobotics/micro-rdk/tree/main/micro-rdk-installer) is a CLI that allows you to flash a build of Micro-RDK, along with your robot's credentials and your wifi information, directly to your ESP32.
```

> Viam provides the [Micro-RDK Installer](https://github.com/viamrobotics/micro-rdk/tree/main/micro-rdk-installer), a CLI that allows a user to flash a build of Micro-RDK, along with their robot's credentials and their wifi information, directly to their ESP32.

Author: Will do, sorry about forgetting to! updated

Select the appropriate pre-built binary for the architecture of your development machine:

{{< tabs name="Download pre-built binaries" >}}
{{% tab name="Linux (x86_64)"%}}

[Download](https://github.com/viamrobotics/micro-rdk/releases/latest/download/micro-rdk-installer-amd64-linux)

{{% /tab %}}
{{% tab name="Linux (Aarch64)" %}}

[Download](https://github.com/viamrobotics/micro-rdk/releases/latest/download/micro-rdk-installer-arm64-linux)

{{% /tab %}}
{{% tab name="MacOS" %}}

[Download](https://github.com/viamrobotics/micro-rdk/releases/latest/download/micro-rdk-installer-macos)

{{% /tab %}}
{{% tab name="Windows" %}}

[Download](https://github.com/viamrobotics/micro-rdk/releases/latest/download/micro-rdk-installer-windows.exe)

{{% /tab %}}
{{< /tabs >}}

### Flash your ESP32 with the Micro RDK Installer

Navigate to [the Viam app](https://app.viam.com) and [add a new robot](/manage/fleet/robots/#add-a-new-robot) in your desired location.

1. Click on the name of the robot to go to the robot's page.
2. Click on the **Setup** tab.
3. Regardless of your operating system, select **Mac** and press the button that appears in Step 2 to download the Viam app config for your robot.

>    Copy the path where this <file>viam.json</file> file was downloaded.

Author: Done

>    In your terminal, cd to the directory when you downloaded your pre-built binary and run the following command to flash the micro-RDK directly to an ESP32 connected to your computer via data cable:

Author: ```suggestion
   In your terminal, cd to the directory when you downloaded your pre-built binary and run the following command to flash the micro-RDK directly to an ESP32 connected to your computer through a data cable:
```

>    In your terminal, cd to the directory when you downloaded your pre-built binary and run the following command to flash the micro-RDK directly to an ESP32 connected to your computer through a data cable:

Author: ```suggestion
   In your terminal, `cd` to the directory where you downloaded your pre-built binary and run the following command to flash the micro-RDK directly to an ESP32 connected to your computer through a data cable:
```

   {{< tabs name="Download pre-built binaries" >}}
   {{% tab name="Linux (x86_64)"%}}

```sh { class="command-line" data-prompt="$"}
./micro-rdk-installer-amd64-linux write-flash --app-config=<your-file-path-to/viam.json>
```

    {{% /tab %}}
    {{% tab name="Linux (Aarch64)" %}}

```sh { class="command-line" data-prompt="$"}

> ./micro-rdk-installer-arm64-linux write-flash --app-config=<your-file-path-to/viam.json>

Author: Sorry, I might be confused , it is different and I think that's supposed to be the case!

> ./micro-rdk-installer-arm64-linux write-flash --app-config=<your-file-path-to/viam.json>

Reviewer 2: I think I misread sorry

```

    {{% /tab %}}
    {{% tab name="MacOS" %}}

```sh { class="command-line" data-prompt="$"}
./micro-rdk-installer-macos write-flash --app-config=<your-file-path-to/viam.json>
```

    {{% /tab %}}
    {{% tab name="Windows" %}}

```sh { class="command-line" data-prompt="$"}
./micro-rdk-installer-windows write-flash --app-config=<your-file-path-to/viam.json>
```

    {{% /tab %}}
    {{< /tabs >}}

    To see the micro-RDK server logs through the serial connection, add `--monitor`.
    If the program cannot auto-detect the serial port to which your ESP32 is connected, you may be prompted to select the correct one among a list.

Go back to your new robot's page on [the Viam app](https://app.viam.com).

> If successful, **Live** should be displayed underneath **Last online**.

Reviewer 2: this would benefit from a screenshot, or from a shorter description:
```suggestion
If successful, your robot will show that it's **Live**.
```

> For more `micro-rdk-installer` CLI usage options, see [GitHub](https://github.com/viamrobotics/micro-rdk/tree/main/micro-rdk-installer).

Author: ```suggestion
For more `micro-rdk-installer` CLI usage options, see [GitHub](https://github.com/viamrobotics/micro-rdk/tree/main/micro-rdk-installer).
```

@gvaradarajan Here I could also document how to write Wi-fi credentials, but I'm not sure what the command for that would look like exactly from readme after `./micro-rdk-installer-amd64-linux write-credentials`. Do you want to document this on docs.viam, and if so do you have an example of providing credentials in a command that I could use?

Author: Ok, sounds good to me!

| ---- | ---- | --------- | ----------- |
| `pins` | object | Required | The {{< glossary_tooltip term_id="pin-number" text="pin number" >}} of any GPIO pins you wish to use as input/output with the [`GPIOPin` API](/program/apis/#gpio-pins). |
| `analogs` | object | Optional | Attributes of any pins that can be used as analog-to-digital converter (ADC) inputs. See [configuration info](/components/board/#analogs). |

### Troubleshooting

#### Linux Port Permissions

If a "Permission Denied" or similar port error occurs, first check the connection of the ESP32 to the machine's USB port.

> If connected and the error persists, run `sudo usermod -a -G dialout $USER` to add the current user to the `dialout` group, then try again.

Reviewer: Add that the user may need to restart the terminal before trying again

Author: done!

| ---- | ---- | --------- | ----------- |
| `pins` | object | Required | The {{< glossary_tooltip term_id="pin-number" text="pin number" >}} of any GPIO pins you wish to use as input/output with the [`GPIOPin` API](/program/apis/#gpio-pins). |
| `analogs` | object | Optional | Attributes of any pins that can be used as analog-to-digital converter (ADC) inputs. See [configuration info](/components/board/#analogs). |

### Troubleshooting

#### Linux Port Permissions

If a "Permission Denied" or similar port error occurs, first check the connection of the ESP32 to the machine's USB port.
If connected and the error persists, run `sudo usermod -a -G dialout $USER` to add the current user to the `dialout` group, then try again.

#### MacOS Executable Permissions

> When using a machine running a version of MacOS, the user will be blocked from running the executable.

Reviewer 2: ```suggestion
When using a machine running a version of MacOS, the user is blocked from running the executable by default.
```

Author: Done
